### ISSUE TITLE
spec: cannot check dynamic array sizes with type assertion/switch

### ISSUE THREAD

**Issue by gjh33** [golang/go#27257]
Please answer these questions before submitting your issue. Thanks!

### What version of Go are you using (`go version`)?
go version go1.10.3

### Does this issue reproduce with the latest release?
Unknown, but likely

### What operating system and processor architecture are you using (`go env`)?
set GOARCH=amd64
set GOEXE=.exe
set GOHOSTARCH=amd64
set GOHOSTOS=windows
set GOOS=windows
set CC=gcc
set CXX=g++
set CGO_ENABLED=1
set CGO_CFLAGS=-g -O2
set CGO_CPPFLAGS=
set CGO_CXXFLAGS=-g -O2
set CGO_FFLAGS=-g -O2
set CGO_LDFLAGS=-g -O2
set PKG_CONFIG=pkg-config

### What did you do?

https://play.golang.org/p/DFS7XKRN-Di

Check an interface to be an array of size x where x is a variable
Can't be done in type switch or type assertion

### What did you expect to see?

Valid type check for the size of the array, Or some way to check if it is an array (without reflection)
to then test the size. Furthermore in my personal case I want to know it's a [x]float32 array. The user
should have the option to pass an array and have me verify it's the correct size for the data it should
represent

### What did you see instead?

prog.go:9:18: non-constant array bound sizeToCheck

I'd argue this is a valuable feature to have, as in my application I'd like the user to be able to pass an
array of fixed size, however that fix size is dependent upon an external variable. In my case a shader parameter is being read from the compiled shader program. That parameter can be an array with ArraySize. So if the user is trying to set a vec2, they should pass an array of [ArraySize * 2]float32. Now it's perfectly valid the user will know the contents of the shader and want to hard code an array for quick debugging and use internally. However my end of the program won't know this ahead of time until it's checked the shader at runtime. Therefore i've ommited array support for my value setting function, but it would be a performant solution for hardcoded default matrices

**Comment by ianlancetaylor** [comment]
That is how the language works today.  I gather that you want to change the language; in that case, I recommend that you develop a proposal for how it should change; see https://golang.org/s/proposal.

People who wrote code that works with arrays of different lengths normally use slices.  It's easy to convert from an array to a slice.  Can you explain why that doesn't work for you?  In what kind of code would you expect to work with arrays of varying sizes?

Also, as you note, you can use the reflect package.  Why is that insufficient?

**Comment by gjh33** [comment]
I'm not looking to use Arrays of variable sizes, my problem involves checking the size of a non variable array that a user may provide, however I need to verify it's size based on a runtime variable.

More specifically: User A writes a shader. This shader has a vec2 array of size 3. This user knows he wrote this file, so rightfully makes a [6]float32 for this variable. This is a valid case for using a constant sized array no? Where he uses his knowledge of the size to make an array whose size won't change. Now it's my turn. My program loads up the shader program, from the shader program it gets the type "vec2" and the ArraySize = 3. When the user now calls my program to SetShaderUniform(data) I would like to ensure the data he's passing me is either:
 1) Array of size 3 * 2
 2) A slice with len 3 * 2
 3) a pointer to the start of a block of memory containing 3 * 2 floats

In regards to the reflect package, I'm concerned about performance since my library is at a low level.

**Comment by ianlancetaylor** [comment]
What if you simply require your user to slice the array by writing `a[:]` when calling your function?

**Comment by josharian** [comment]
Tangentially related: #395

**Comment by ianlancetaylor** [comment]
@gjh33 Right now array types are simple: the length must be known at compile time.  You are suggesting that we make them more complicated: at least in some cases, the length may be known only at run time.  Making the language more complicated is a cost: there are more rules that people need to remember.  Any cost needs a corresponding benefit.  I was asking questions to try to better understand the benefit.  At this point, the benefit seems small.  It's not obvious to me that it is worth the cost.

**Comment by gjh33** [comment]
Maybe that's the case, I just felt that the optimizations and benefits of arrays' size being known at compile time is something I'd want users of my library to have the option to. And I really like the idea of the compile time optimizations that can be made around arrays, but for now their use seems extremely limited to me if I can't check the size. That being said I'm willing to admit there could be costs larger than all the potential use cases that are opened up, which is why I opened this bug report (which is really a feature request) was to get the opinion of someone such as yourself. I think at this point I've stated my case for it and if you feel there are costs that outweigh the potential increased use cases for compile time optimized arrays, then I yield and this thread can be closed.
